store/s3storeKMS.go

```diff
@@ -0,0 +1,478 @@
+package store
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sort"
+	"time"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/pkg/errors"
+)
+
+// latest is used to keep a single object in s3 with all of the
+// most recent values for the given service's secrets.  Keeping this
+// in a single s3 object allows us to use a single s3 GetObject
+// for ListRaw (and thus chamber exec).
+type LatestIndexFile struct {
+	Latest map[string]LatestValue `json:"latest"`
+}
+
+type LatestValue struct {
+	Version   int       `json:"version"`
+	Value     string    `json:"value"`
+	KMSAlias  string    `json:"KMSAlias"`
+}
+
+var _ Store = &S3KMSStore{}
+
+type S3KMSStore struct {
+	svc    s3iface.S3API
+	stsSvc *sts.STS
+	bucket string
+}
+
+// Deprecated; use NewS3KMSStoreWithBucket instead
+func NewS3KMSStore(numRetries int) (*S3KMSStore, error) {
+	bucket, ok := os.LookupEnv(BucketEnvVar)
+	if !ok {
+		return nil, fmt.Errorf("Must set %s for s3 backend", BucketEnvVar)
+	}
+
+	return NewS3KMSStoreWithBucket(numRetries, bucket)
+}
+
+func(s *S3KMSStore) KMSKey() string {
+	fromEnv, ok := os.LookupEnv("CHAMBER_KMS_KEY_ALIAS")
+	if !ok {
+		return DefaultKeyID
+	}
+	if !strings.HasPrefix(fromEnv, "alias/") {
+		return fmt.Sprintf("alias/%s", fromEnv)
+	}
+
+	return fromEnv
+}
+
+func NewS3KMSStoreWithBucket(numRetries int, bucket string) (*S3KMSStore, error) {
+	session, region, err := getSession(numRetries)
+	if err != nil {
+		return nil, err
+	}
+
+	svc := s3.New(session, &aws.Config{
+		MaxRetries: aws.Int(numRetries),
+		Region:     region,
+	})
+
+	stsSvc := sts.New(session, &aws.Config{
+		MaxRetries: aws.Int(numRetries),
+		Region:     region,
+	})
+
+	return &S3KMSStore{
+		svc:    svc,
+		stsSvc: stsSvc,
+		bucket: bucket,
+	}, nil
+}
+
+func (s *S3KMSStore) Write(id SecretId, value string) error {
+	index, err := s.readLatest(id.Service)
+	if err != nil {
+		return err
+	}
+
+	if val, ok := index.Latest[id.Key]; ok {
+		if val.KMSAlias != s.KMSKey() {
+			return errors.New(fmt.Sprintf("You're attempting to write a Chamber key with a different KMS Key to the one it currently uses. Instead delete the existing chamber key using the current KMS key (%s) and then write the secret using the new KMS Key.", val.KMSAlias))
```

Comment:
I'm not sure I understand this requirement. What if you've lost access to the old key?

Reply:
> What if you've lost access to the old key?

You wouldn't be able to read/write it anyway without access to the old key.

> I'm not sure I understand this requirement.

Let KMSKeyA be a KMS Key
Let KMSKeyB be a KMS Key

Let SecretA be a Chamber Secret
Let SecretB be a Chamber Secret

Let ChamberOfSecrets be a Chamber Service

User A has access to KMSKeyA & KMSKeyB
User B has access to only KMSKeyA
User C has access to only KMSKeyB

User A writes SecretA with value ABC to ChamberOfSecrets with KMSKeyA
- This creates `kms__KMSKeyA__latest.json` and stores `{ SecretA: "ABC" }`

UserB writes SecretB with value 123 to ChamberOfSecrets with KMSKeyA
- This uses `kms__KMSKeyA__latest.json` and stores `{ SecretA: "ABC", SecretB: 123 }`

UserA attempts to write SecretA with value DEF to ChamberOfSecrets with KMSKeyB
- This uses `kms__KMSKeyB__latest.json` and stores `{ SecretA: "DEF" }`
- `kms__KMSKeyA__latest.json` *still has the value* `{ SecretA: "ABC", SecretB: 123 }`

User B can now see the value of Secret A as "ABC" and User B can see the value as "DEF".

In order for Chamber to migrate the secret over it needs to delete SecretA from `kms__KMSKeyB__latest.json` and write it to `kms__KMSKeyA__latest.json` but the operation that you first asked for was `chamber write`. I don't want a write command to actually delete any data. I want to explicitly delete and then write.